pkg/datapath/linux/ipsec/ipsec_linux.go

```diff
@@ -752,18 +752,67 @@ func ipsecDeleteXfrmState(nodeID uint16) error {
 		return err
 	}
 
-	errs := resiliency.NewErrorSet(fmt.Sprintf("failed to delete node (%d) xfrm states", nodeID), len(xfrmStateList))
+	xfrmStatesToDelete := []netlink.XfrmState{}
 	for _, s := range xfrmStateList {
 		if matchesOnNodeID(s.Mark) && ipsec.GetNodeIDFromXfrmMark(s.Mark) == nodeID {
-			if err := netlink.XfrmStateDel(&s); err != nil {
-				errs.Add(fmt.Errorf("failed to delete xfrm state (%s): %w", s.String(), err))
-			}
+			xfrmStatesToDelete = append(xfrmStatesToDelete, s)
+		}
+	}
+
+	errs := resiliency.NewErrorSet(fmt.Sprintf("failed to delete node (%d) xfrm states", nodeID), len(xfrmStateList))
+	for _, s := range xfrmStatesToDelete {
+		if err := safeDeleteXfrmState(&s, xfrmStateList); err != nil {
+			errs.Add(fmt.Errorf("failed to delete xfrm state (%s): %w", s.String(), err))
 		}
 	}
 
 	return errs.Error()
 }
 
+// safeDeleteXfrmState deletes the given XFRM state. Specifically, if the
+// state is to catch ingress traffic marked with nodeID (0xXXXX0d00), we
+// temporarily remove the old XFRM state that matches 0xd00/0xf00. This is to
+// workaround a kernel issue that prevents us from deleting a specific XFRM
+// state (e.g. catching 0xXXXX0d00/0xffff0f00) when there is also a general
+// xfrm state (e.g. catching 0xd00/0xf00). When both XFRM states coexist,
+// kernel deletes the general XFRM state instead of the specific one, even if
+// the deleting request is for the specific one.
+func safeDeleteXfrmState(state *netlink.XfrmState, stateList []netlink.XfrmState) (err error) {
+	if getDirFromXfrmMark(state.Mark) == dirIngress && ipsec.GetNodeIDFromXfrmMark(state.Mark) != 0 {
+		oldXFRMInMark := &netlink.XfrmMark{
+			Value: linux_defaults.RouteMarkDecrypt,
+			Mask:  linux_defaults.IPsecMarkBitMask,
+		}
+
+		errs := resiliency.NewErrorSet("failed to delete old xfrm states", len(stateList))
+
+		scopedLog := log.WithFields(logrus.Fields{
+			logfields.SPI:              state.Spi,
+			logfields.SourceIP:         state.Src,
+			logfields.DestinationIP:    state.Dst,
+			logfields.TrafficDirection: getDirFromXfrmMark(state.Mark),
+			logfields.NodeID:           getNodeIDAsHexFromXfrmMark(state.Mark),
+		})
+
+		for _, s := range stateList {
+			if s.Spi == state.Spi && xfrmIPEqual(s.Dst, state.Dst) && xfrmMarkEqual(s.Mark, oldXFRMInMark) {
+				err, deferFn := xfrmTemporarilyRemoveState(scopedLog, s, string(dirIngress))
+				if err != nil {
+					errs.Add(fmt.Errorf("Failed to remove old XFRM %s state %s: %w", string(dirIngress), s.String(), err))
+				} else {
+					defer deferFn()
+				}
+			}
+		}
```

Comment:
Can we maybe avoid this second walk of the `stateList`, and instead have the caller already do the comparison?

Reply:
looks good now, ty!